r/R/query-engine.R

```diff
@@ -384,3 +397,65 @@ needs_projection <- function(projection, schema) {
     !identical(field_names, names(projection)) || # Any fields are renamed
     !identical(field_names, names(schema)) # The fields are reordered
 }
+
+#' Create projection needed to coalesce join keys after a full outer join
+#'
+#' @examples
+#' test_join <- list(
+#'   type = JoinType$FULL_OUTER,
+#'   right_data = arrow_table(x = 1, y = 2, z = "x"),
+#'   by = c("x", "y"),
+#'   suffix = c(".x", ".y"),
+#'   keep = FALSE
+#' )
+#' post_join_projection(c("value", "x", "y", "z"), test_join)
+#'
+#' @noRd
+post_join_projection <- function(left_names, join_config) {
+  # Need to coalesce the key columns
+  right_names <- names(join_config$right_data)
+  # Collect mapping of which columns on left need to be coalesced with which
+  # column on the right side.
+  coalesce_targets <- data.frame(
```

Comment:
Rather than a data.frame, would a named list or character vector be a simpler map to work with?

Reply:
I'm not sure how that would work. If there's something like a Python dictionary that can represent `Dict[int, int]` that would be nice. But it would only eliminate one line I think.